Fix sqlite concurrent access bug

### Fix sqlite concurrent access bug
Issue was found on random travis run https://travis-ci.com/algorand/go-algorand/jobs/290721815; It is likely to be triggered by the recent generateGenesisFiles optimization, but the optimization *should* be harmless.

The travis log for the error goes as follows:
```
[./scripts/check_deps.sh] Required dependencies installed.
2380go install  --tags "sqlite_unlock_notify sqlite_omit_load_extension osusergo netgo static_build" -buildmode pie -ldflags="-X github.com/algorand/go-algorand/config.BuildNumber=64022 -X github.com/algorand/go-algorand/config.CommitHash=897c480a -X github.com/algorand/go-algorand/config.Branch=master -X github.com/algorand/go-algorand/config.DefaultDeadlock=enable -extldflags \"-static-libstdc++ -static-libgcc -static\" -X github.com/algorand/go-algorand/config.Channel=master" ./...
2381/home/travis/gopath/bin/genesis -q -n testnet -c gen/testnet.json -d gen/testnet
2382./scripts/dump_genesis.sh gen/testnet/genesis.json > gen/testnet/genesis.dump
2383/home/travis/gopath/bin/genesis -q -n devnet -c gen/devnet.json -d gen/devnet
2384fatal error: unexpected signal during runtime execution
2385[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x0]
2386
2387runtime stack:
2388runtime.throw(0xdf6022, 0x2a)
2389	/home/travis/.gimme/versions/go1.12.linux.amd64/src/runtime/panic.go:617 +0x74
2390runtime.sigpanic()
2391	/home/travis/.gimme/versions/go1.12.linux.amd64/src/runtime/signal_unix.go:374 +0x4ad
2392
2393goroutine 11 [syscall]:
2394runtime.cgocall(0xb883a0, 0xc0000ccbb8, 0xc0000100d0)
2395	/home/travis/.gimme/versions/go1.12.linux.amd64/src/runtime/cgocall.go:128 +0x5d fp=0xc0000ccb88 sp=0xc0000ccb50 pc=0x40b02d
2396github.com/mattn/go-sqlite3._Cfunc__sqlite3_open_v2(0x362cf90, 0xc0000100d0, 0xc000010006, 0x0, 0x0)
2397	_cgo_gotypes.go:290 +0x4f fp=0xc0000ccbb8 sp=0xc0000ccb88 pc=0xb37d1f
2398github.com/mattn/go-sqlite3.(*SQLiteDriver).Open.func2(0x362cf90, 0xc0000100d0, 0x10000, 0x1)
2399	/home/travis/gopath/pkg/mod/github.com/mattn/go-sqlite3@v1.10.0/sqlite3.go:1341 +0xbd fp=0xc0000ccbf0 sp=0xc0000ccbb8 pc=0xb4ef3d
2400github.com/mattn/go-sqlite3.(*SQLiteDriver).Open(0xc00000e140, 0xc000034240, 0x56, 0x0, 0x0, 0x0, 0x0)
2401	/home/travis/gopath/pkg/mod/github.com/mattn/go-sqlite3@v1.10.0/sqlite3.go:1341 +0x1133 fp=0xc0000cd478 sp=0xc0000ccbf0 pc=0xb435a3
2402database/sql.dsnConnector.Connect(...)
2403	/home/travis/.gimme/versions/go1.12.linux.amd64/src/database/sql/sql.go:636
2404database/sql.(*dsnConnector).Connect(0xc00000e460, 0x15c20c0, 0xc0000ac110, 0x0, 0x0, 0x0, 0x40)
2405	<autogenerated>:1 +0x56 fp=0xc0000cd4c0 sp=0xc0000cd478 pc=0x8e6836
2406database/sql.(*DB).conn(0xc00021c180, 0x15c20c0, 0xc0000ac110, 0x1, 0x45207a, 0xbb688fe76e, 0xc011604648)
2407	/home/travis/.gimme/versions/go1.12.linux.amd64/src/database/sql/sql.go:1176 +0x12d fp=0xc0000cd668 sp=0xc0000cd4c0 pc=0x8da42d
2408database/sql.(*DB).Conn(0xc00021c180, 0x15c20c0, 0xc0000ac110, 0x10, 0x14b38c0, 0xc00021c180)
2409	/home/travis/.gimme/versions/go1.12.linux.amd64/src/database/sql/sql.go:1695 +0x84 fp=0xc0000cd6e8 sp=0xc0000cd668 pc=0x8de1c4
2410github.com/algorand/go-algorand/util/db.Accessor.Atomic(0xc00021c180, 0xc000040301, 0x0, 0x0, 0xc000040300, 0x0, 0x0, 0x0, 0x0, 0x0)
2411	/home/travis/gopath/src/github.com/algorand/go-algorand/util/db/dbutil.go:193 +0xd3 fp=0xc0000cd7c0 sp=0xc0000cd6e8 pc=0xb534b3
2412github.com/algorand/go-algorand/data/account.RestoreRoot(0xc00021c180, 0x1, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x24, ...)
2413	/home/travis/gopath/src/github.com/algorand/go-algorand/data/account/account.go:90 +0xe1 fp=0xc0000cd848 sp=0xc0000cd7c0 pc=0xb79c21
2414github.com/algorand/go-algorand/gen.loadRootKey(0xc000026880, 0x1a, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
2415	/home/travis/gopath/src/github.com/algorand/go-algorand/gen/generate.go:341 +0x26b fp=0xc0000cd950 sp=0xc0000cd848 pc=0xb84a0b
2416github.com/algorand/go-algorand/gen.generateGenesisFiles.func2()
2417	/home/travis/gopath/src/github.com/algorand/go-algorand/gen/generate.go:149 +0x50c fp=0xc0000cdfe0 sp=0xc0000cd950 pc=0xb85a0c
2418runtime.goexit()
2419	/home/travis/.gimme/versions/go1.12.linux.amd64/src/runtime/asm_amd64.s:1337 +0x1 fp=0xc0000cdfe8 sp=0xc0000cdfe0 pc=0x460421
2420created by github.com/algorand/go-algorand/gen.generateGenesisFiles
2421	/home/travis/gopath/src/github.com/algorand/go-algorand/gen/generate.go:252 +0x410
2422
2423goroutine 1 [semacquire]:
2424sync.runtime_Semacquire(0xc000028678)
2425	/home/travis/.gimme/versions/go1.12.linux.amd64/src/runtime/sema.go:56 +0x3b
2426sync.(*WaitGroup).Wait(0xc000028670)
2427	/home/travis/.gimme/versions/go1.12.linux.amd64/src/sync/waitgroup.go:130 +0x67
2428github.com/algorand/go-algorand/gen.generateGenesisFiles(0x7ffee8242910, 0xa, 0xc0000320f0, 0x4e, 0x2710, 0x2328, 0x2710, 0x0, 0x0, 0x80, ...)
2429	/home/travis/gopath/src/github.com/algorand/go-algorand/gen/generate.go:256 +0x436
2430github.com/algorand/go-algorand/gen.GenerateGenesisFiles(0x7ffee82428f3, 0x6, 0xc000028568, 0x2, 0xc0000320f0, 0x4e, 0x0, 0x2dc6c0, 0x0, 0xc000059180, ...)
2431	/home/travis/gopath/src/github.com/algorand/go-algorand/gen/generate.go:106 +0x5df
2432main.main()
2433	/home/travis/gopath/src/github.com/algorand/go-algorand/cmd/genesis/newgenesis.go:57 +0x1b1
2434
2435goroutine 8 [runnable]:
2436github.com/algorand/go-algorand/gen.generateGenesisFiles.func2()
2437	/home/travis/gopath/src/github.com/algorand/go-algorand/gen/generate.go:133
2438created by github.com/algorand/go-algorand/gen.generateGenesisFiles
2439	/home/travis/gopath/src/github.com/algorand/go-algorand/gen/generate.go:252 +0x410
2440
2441goroutine 9 [runnable]:
2442github.com/mattn/go-sqlite3._Cfunc__sqlite3_open_v2(0x7f53e8000b20, 0xc0000b0548, 0xc000010006, 0x0, 0x0)
2443	_cgo_gotypes.go:290 +0x4f
2444github.com/mattn/go-sqlite3.(*SQLiteDriver).Open.func2(0x7f53e8000b20, 0xc0000b0548, 0x7f5300010000, 0x1)
2445	/home/travis/gopath/pkg/mod/github.com/mattn/go-sqlite3@v1.10.0/sqlite3.go:1341 +0xbd
2446github.com/mattn/go-sqlite3.(*SQLiteDriver).Open(0xc00000e140, 0xc000086720, 0x57, 0x0, 0x0, 0x0, 0x0)
2447	/home/travis/gopath/pkg/mod/github.com/mattn/go-sqlite3@v1.10.0/sqlite3.go:1341 +0x1133
2448database/sql.dsnConnector.Connect(...)
2449	/home/travis/.gimme/versions/go1.12.linux.amd64/src/database/sql/sql.go:636
2450database/sql.(*DB).conn(0xc000112180, 0x15c20c0, 0xc0000ac110, 0x1, 0x45207a, 0xbb688fe773, 0xc01160464a)
2451	/home/travis/.gimme/versions/go1.12.linux.amd64/src/database/sql/sql.go:1176 +0x12d
2452database/sql.(*DB).Conn(0xc000112180, 0x15c20c0, 0xc0000ac110, 0x10, 0x14b38c0, 0xc000112180)
2453	/home/travis/.gimme/versions/go1.12.linux.amd64/src/database/sql/sql.go:1695 +0x84
2454github.com/algorand/go-algorand/util/db.Accessor.Atomic(0xc000112180, 0xc000194b01, 0x0, 0x0, 0xc000194b10, 0x0, 0x0, 0x0, 0x0, 0x0)
2455	/home/travis/gopath/src/github.com/algorand/go-algorand/util/db/dbutil.go:193 +0xd3
2456github.com/algorand/go-algorand/data/account.RestoreRoot(0xc000112180, 0x1, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x25, ...)
2457	/home/travis/gopath/src/github.com/algorand/go-algorand/data/account/account.go:90 +0xe1
2458github.com/algorand/go-algorand/gen.loadRootKey(0xc0000ae960, 0x1b, 0x0, 0x0, 0x0, 0x0, 0x0, 0x14d3860, 0x0, 0x0, ...)
2459	/home/travis/gopath/src/github.com/algorand/go-algorand/gen/generate.go:341 +0x26b
2460github.com/algorand/go-algorand/gen.generateGenesisFiles.func2()
2461	/home/travis/gopath/src/github.com/algorand/go-algorand/gen/generate.go:149 +0x50c
2462created by github.com/algorand/go-algorand/gen.generateGenesisFiles
2463	/home/travis/gopath/src/github.com/algorand/go-algorand/gen/generate.go:252 +0x410
2464
2465goroutine 10 [runnable]:
2466github.com/algorand/go-algorand/gen.generateGenesisFiles.func2()
2467	/home/travis/gopath/src/github.com/algorand/go-algorand/gen/generate.go:133
2468created by github.com/algorand/go-algorand/gen.generateGenesisFiles
2469	/home/travis/gopath/src/github.com/algorand/go-algorand/gen/generate.go:252 +0x410
2470
2471goroutine 12 [runnable]:
2472database/sql.(*DB).connectionOpener(0xc00021c180, 0x15c2080, 0xc00007a300)
2473	/home/travis/.gimme/versions/go1.12.linux.amd64/src/database/sql/sql.go:998
2474created by database/sql.OpenDB
2475	/home/travis/.gimme/versions/go1.12.linux.amd64/src/database/sql/sql.go:670 +0x160
2476
2477goroutine 28 [runnable]:
2478database/sql.(*DB).connectionOpener(0xc000112180, 0x15c2080, 0xc0001a84c0)
2479	/home/travis/.gimme/versions/go1.12.linux.amd64/src/database/sql/sql.go:998
2480created by database/sql.OpenDB
2481	/home/travis/.gimme/versions/go1.12.linux.amd64/src/database/sql/sql.go:670 +0x160
2482
2483goroutine 29 [runnable]:
2484database/sql.(*DB).connectionResetter(0xc000112180, 0x15c2080, 0xc0001a84c0)
2485	/home/travis/.gimme/versions/go1.12.linux.amd64/src/database/sql/sql.go:1011
2486created by database/sql.OpenDB
2487	/home/travis/.gimme/versions/go1.12.linux.amd64/src/database/sql/sql.go:671 +0x196
2488
2489goroutine 13 [runnable]:
2490database/sql.(*DB).connectionResetter(0xc00021c180, 0x15c2080, 0xc00007a300)
2491	/home/travis/.gimme/versions/go1.12.linux.amd64/src/database/sql/sql.go:1011
2492created by database/sql.OpenDB
2493	/home/travis/.gimme/versions/go1.12.linux.amd64/src/database/sql/sql.go:671 +0x196
2494Makefile:247: recipe for target 'gen/devnet/genesis.json' failed
2495make: *** [gen/devnet/genesis.json] Error 2
2496/home/travis/gopath/src/github.com/algorand/go-algorand/scripts/travis/travis_wait.sh: line 34: 10784 Terminated              travis_jigger $! $timeout $cmd
2497The command "scripts/travis/build_test.sh" exited with 2.
cache.2
2498store build cache
```

--- a/gen/generate.go
+++ b/gen/generate.go
@@ -21,7 +21,7 @@ import (
 	"io/ioutil"
 	"os"
 	"path/filepath"
-	//"runtime"
+	"runtime"
 	"sort"
 	"sync"
 	"sync/atomic"
@@ -125,9 +125,7 @@ func generateGenesisFiles(outDir string, protoVersion protocol.ConsensusVersion,
 
 	pendingWallets := make(chan genesisAllocation, len(allocation))
 
-	// temporary disable the concurrent execution, as it seems to create an issue with sqlite concurrently model.
-	// this should be removed once the undelying issue is resolved.
-	concurrentWalletGenerators := int(1) // runtime.NumCPU() * 2
+	concurrentWalletGenerators := runtime.NumCPU() * 2
 	errorsChannel := make(chan error, concurrentWalletGenerators)
 	verbosedOutput := make(chan string)
 	var creatingWalletsWaitGroup sync.WaitGroup

--- a/gen/generate_test.go
+++ b/gen/generate_test.go
@@ -1,0 +1,102 @@
+// Copyright (C) 2019-2020 Algorand, Inc.
+// This file is part of go-algorand
+//
+// go-algorand is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, either version 3 of the
+// License, or (at your option) any later version.
+//
+// go-algorand is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with go-algorand.  If not, see <https://www.gnu.org/licenses/>.
+
+package gen
+
+import (
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"sync"
+	"testing"
+
+	"github.com/algorand/go-algorand/data/account"
+	"github.com/algorand/go-algorand/util/db"
+
+	"github.com/stretchr/testify/require"
+)
+
+func TestLoadMultiRootKeyConcurrent(t *testing.T) {
+	t.Skip() // skip in auto-test mode
+	a := require.New(t)
+	tempDir, err := ioutil.TempDir("", "loadkey-test-")
+	a.NoError(err)
+	defer os.RemoveAll(tempDir)
+
+	const numThreads = 100
+	var wg sync.WaitGroup
+	wg.Add(numThreads)
+
+	for i := 0; i < numThreads; i++ {
+		go func(idx int) {
+			defer wg.Done()
+			wallet := filepath.Join(tempDir, fmt.Sprintf("wallet%d", idx+1))
+			rootDB, err := db.MakeErasableAccessor(wallet)
+			defer rootDB.Close()
+			a.NoError(err)
+			_, err = account.GenerateRoot(rootDB)
+			a.NoError(err)
+		}(i)
+	}
+
+	wg.Wait()
+
+	for r := 0; r < 1000; r++ {
+		var wg sync.WaitGroup
+		wg.Add(numThreads)
+		for i := 0; i < numThreads; i++ {
+			go func(idx int) {
+				defer wg.Done()
+				wallet := filepath.Join(tempDir, fmt.Sprintf("wallet%d", idx+1))
+				_, db, err := loadRootKey(wallet)
+				a.NoError(err)
+				db.Close()
+			}(i)
+		}
+		wg.Wait()
+	}
+}
+
+func TestLoadSingleRootKeyConcurrent(t *testing.T) {
+	t.Skip() // skip in auto-test mode
+	a := require.New(t)
+	tempDir, err := ioutil.TempDir("", "loadkey-test-")
+	a.NoError(err)
+	defer os.RemoveAll(tempDir)
+
+	wallet := filepath.Join(tempDir, "wallet1")
+	rootDB, err := db.MakeErasableAccessor(wallet)
+	a.NoError(err)
+	_, err = account.GenerateRoot(rootDB)
+	rootDB.Close()
+	a.NoError(err)
+
+	const numThreads = 10000
+	var wg sync.WaitGroup
+	wg.Add(numThreads)
+
+	for i := 0; i < numThreads; i++ {
+		go func(idx int) {
+			defer wg.Done()
+			wallet := filepath.Join(tempDir, "wallet1")
+			_, db, err := loadRootKey(wallet)
+			a.NoError(err)
+			db.Close()
+		}(i)
+	}
+	wg.Wait()
+}

--- a/util/db/dbutil.go
+++ b/util/db/dbutil.go
@@ -26,6 +26,8 @@ import (
 	"fmt"
 	"reflect"
 	"runtime"
+	"strings"
+	"sync"
 	"time"
 
 	"github.com/mattn/go-sqlite3"
@@ -48,6 +50,7 @@ const busy = 1000
 // connection.  For now, it's just an optional "PRAGMA fullfsync=true" on
 // MacOSX.
 var initStatements []string
+var sqliteInitOnce sync.Once
 
 // An Accessor manages a sqlite database handle and any outstanding batching operations.
 type Accessor struct {
@@ -58,30 +61,48 @@ type Accessor struct {
 
 // MakeAccessor creates a new Accessor.
 func MakeAccessor(dbfilename string, readOnly bool, inMemory bool) (Accessor, error) {
-	var db Accessor
-	db.readOnly = readOnly
-
-	var err error
-	db.Handle, err = sql.Open("sqlite3", URI(dbfilename, readOnly, inMemory)+"&_journal_mode=wal")
-
-	if err == nil {
-		err = db.runInitStatements()
-	}
-
-	return db, err
+	return makeAccessorImpl(dbfilename, readOnly, inMemory, []string{"_journal_mode=wal"})
 }
 
 // MakeErasableAccessor creates a new Accessor with the secure_delete pragma set;
 // see https://www.sqlite.org/pragma.html#pragma_secure_delete
 // It is not read-only and not in-memory (otherwise, erasability doesn't matter)
 func MakeErasableAccessor(dbfilename string) (Accessor, error) {
-	var db Accessor
-	db.readOnly = false
+	return makeAccessorImpl(dbfilename, false, false, []string{"_secure_delete=on"})
+}
 
+func makeAccessorImpl(dbfilename string, readOnly bool, inMemory bool, params []string) (Accessor, error) {
+	var db Accessor
+	db.readOnly = readOnly
+
+	// SQLite3 driver we use (mattn/go-sqlite3) does not implement driver.DriverContext interface
+	// that forces sql.Open calling sql.OpenDB and return a struct without any touches to the underlying driver.
+	// Because of that SQLite library is not initialized until the very first call of sqlite3_open_v2 that happens
+	// in sql.DB.conn. SQLite initialization is not thread-safe on date of writing (2/27/2020) and
+	// mattn/go-sqlite3 has no special code to handle this case.
+	// Solution is to create a connection using a safe synchronization barrier right here.
+	// The connection goes to a connection pool inside Go's sql package and will be re-used when needed.
+	// See https://github.com/algorand/go-algorand/issues/846 for more details.
 	var err error
-	db.Handle, err = sql.Open("sqlite3", URI(dbfilename, false, false)+"&_secure_delete=on")
+	db.Handle, err = sql.Open("sqlite3", URI(dbfilename, readOnly, inMemory)+"&"+strings.Join(params, "&"))
 
 	if err == nil {
+		// create a connection to safely initialize SQLite once
+		initFn := func() {
+			var conn *sql.Conn
+			if conn, err = db.Handle.Conn(context.Background()); err != nil {
+				db.Close()
+				return
+			}
+			if err = conn.Close(); err != nil {
+				db.Close()
+			}
+		}
+		sqliteInitOnce.Do(initFn)
+		if err != nil {
+			// init failed, db closed and err is set
+			return db, err
+		}
 		err = db.runInitStatements()
 	}
 
@@ -93,8 +114,7 @@ func (db *Accessor) runInitStatements() error {
 	for _, stmt := range initStatements {
 		_, err := db.Handle.Exec(stmt)
 		if err != nil {
-			db.Handle.Close()
-			db.Handle = nil
+			db.Close()
 			return err
 		}
 	}
@@ -169,7 +189,7 @@ func (db *Accessor) getDecoratedLogger(fn idemFn, extras ...interface{}) logging
 
 // Atomic executes a piece of code with respect to the database atomically.
 // For transactions where readOnly is false, sync determines whether or not to wait for the result.
-func (db Accessor) Atomic(fn idemFn, extras ...interface{}) (err error) {
+func (db *Accessor) Atomic(fn idemFn, extras ...interface{}) (err error) {
 	start := time.Now()
 
 	// note that the sql library will drop panics inside an active transaction

